network/dnsmasq.go

```diff
@@ -0,0 +1,123 @@
+package network
+
+import (
+	"fmt"
+	"html/template"
+	"os"
+	"os/exec"
+
+	"github.com/adrg/xdg"
+	"github.com/coreos/go-iptables/iptables"
+)
+
+func InstallDnsmasq(name string) error {
+	dnsmasqPath, err := exec.LookPath("dnsmasq")
+	if err != nil {
+		return err
+	}
+
+	pidFile, err := xdg.RuntimeFile("dagger/net/" + name + "/dnsmasq.pid")
+	if err != nil {
+		return err
+	}
+
+	hostsFile, err := xdg.RuntimeFile("dagger/net/" + name + "/hosts")
+	if err != nil {
+		return err
+	}
+
+	upstreamResolvFile, err := xdg.RuntimeFile("dagger/net/" + name + "/resolv.conf.upstream")
+	if err != nil {
+		return err
+	}
+
+	config := dnsmasqConfig{
+		Domain:             name + ".local",
+		NetworkInterface:   name + "0",
+		PidFile:            pidFile,
+		AddnHostsFile:      hostsFile,
+		UpstreamResolvFile: upstreamResolvFile,
+	}
+
+	if err := setupIPTables(config.NetworkInterface); err != nil {
+		return fmt.Errorf("setup iptables: %w", err)
+	}
+
+	dnsmasqConfigFile, err := xdg.RuntimeFile("dagger/net/" + name + "/dnsmasq.conf")
+	if err != nil {
+		return err
+	}
+
+	if err := writeDnsmasqConfig(dnsmasqConfigFile, config); err != nil {
+		return fmt.Errorf("write dnsmasq.conf: %w", err)
+	}
+
+	dnsmasq := exec.Command(dnsmasqPath, "-u", "root", "--conf-file="+dnsmasqConfigFile)
+
+	if b, err := dnsmasq.CombinedOutput(); err != nil {
+		return fmt.Errorf("start dnsmasq: %w; output:\n%s", err, string(b))
+	}
+
+	return nil
+}
+
+func writeDnsmasqConfig(dnsmasqConfigFile string, config dnsmasqConfig) error {
+	conf, err := os.Create(dnsmasqConfigFile)
+	if err != nil {
+		return err
+	}
+
+	defer conf.Close()
+
+	tmpl, err := template.New("dnsmasq.conf").Parse(dnsmasqTemplate)
+	if err != nil {
+		return err
+	}
+
+	if err := tmpl.Execute(conf, config); err != nil {
+		return err
+	}
+
+	return conf.Close()
+}
+
+const dnsmasqTemplate = `## WARNING: THIS IS AN AUTOGENERATED FILE
+## AND SHOULD NOT BE EDITED MANUALLY AS IT
+## LIKELY TO AUTOMATICALLY BE REPLACED.
+strict-order
+local=/{{.Domain}}/
+domain={{.Domain}}
+expand-hosts
+pid-file={{.PidFile}}
+except-interface=lo
+bind-dynamic
+no-hosts
+interface={{.NetworkInterface}}
+addn-hosts={{.AddnHostsFile}}
+resolv-file={{.UpstreamResolvFile}}
+`
+
+type dnsmasqConfig struct {
+	Domain             string
+	NetworkInterface   string
+	PidFile            string
+	AddnHostsFile      string
+	UpstreamResolvFile string
+}
+
+func setupIPTables(iface string) error {
+	ip, err := iptables.New()
+	if err != nil {
+		return err
+	}
+	args := []string{"-i", iface, "-p", "udp", "-m", "udp", "--dport", "53", "-j", "ACCEPT"}
```

Comment:
Honestly not even sure myself. It was there in the original `dnsname` plugin code, confusingly tied to the `dnsmasq.conf` setup, so I kept it: https://github.com/dagger/dagger/pull/4711/files#diff-75e7590bb9ac9d43bb60200ec283d75ac9d28e5ade434a38ea8a6342b8f6956cL60

I can try removing it and see if anything breaks. :man_shrugging: Maybe the `dnsname` plugin had to run in certain places that needed this? It's been there since the original commit so there's no insightful 'blame' for it.

Reply:
(Also I totally had ChatGPT explain what this command did a while back, which it did pretty well!)

Reply:
> I can try removing it and see if anything breaks. 🤷‍♂️ Maybe the dnsname plugin had to run in certain places that needed this? It's been there since the original commit so there's no insightful 'blame' for it.

Cool yeah may as well cut the cargo culting cord at this point I suppose. The less manual fiddling with iptables the better probably